Feature: extend the partition algorithm to handle weights

dutkalex opened the pull request:
**_Describe your changes here:_**
This PR makes it possible to define element weights for the partitioning algorithm. This is achieved via a callback pointer. In the null case, the old unweighted algorithm is used. ~This introduces one breaking change in the public API, in the `t8_forest_set_partition` function signature. All the client code in the repo has been updated, but this will require an update of external user code: `t8_forest_set_partition(forest, set_from, set_for_coarsening)` => `t8_forest_set_partition(forest, set_from, set_for_coarsening, nullptr)`~

*Design tradeoffs:*
- ~The new algorithm replaces the old one, even if no weight function is provided. Keeping the old algorithm around could possibly result in slightly better performance since it requires less work, but this would come at the cost of increased maintenance and the need for duplicated testing to cover both code paths. Given that the partition algorithm is not a performance bottleneck in practice, the single code path solution was preferred.~ The final design includes the unweighted case as a fast "early exit" case.
- In the general case, a single pass algorithm is not possible because the total weight of the forest wan't be known _a priori_ and must be computed (first traversal of the forest) to then be able to define the new partition boundaries (second traversal). This means that each element weight is computed twice with the current implementation. If the weight function is expensive to evaluate, this could cause performance problems. There are basically two ways to work around this: the results could either be cached during the first evaluation (this means that we need to allocate a temporary array), or the weights could alternatively be computed and stored in an array ahead-of-time by the user, before calling the partition function. In the second case the user then simply provides a weight function that looks up the precomputed weights. This second option was preferred because it gives the user greater flexibility, and because it offers optimal performance in the case where the weight function is cheap to evaluate which is probably the default case (_"Don't pay for what you don't use"_ principle).

**_All these boxes must be checked by the AUTHOR before requesting review:_**
- [x] The PR is *small enough* to be reviewed easily. If not, consider splitting up the changes in multiple PRs.
- [x] The title starts with one of the following prefixes: `Documentation:`, `Bugfix:`, `Feature:`, `Improvement:` or `Other:`.
- [ ] If the PR is related to an issue, make sure to link it.
- [ ] The author made sure that, as a reviewer, he/she would check all boxes below.

**_All these boxes must be checked by the REVIEWERS before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is *fully understood, bug free, well-documented and well-structured*.
#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually.
- [ ] The code follows the [t8code coding guidelines](https://github.com/DLR-AMR/t8code/wiki/Coding-Guideline).
- [ ] New source/header files are properly added to the CMake files.
- [ ] The code is well documented. In particular, all function declarations, structs/classes and their members have a proper doxygen documentation.
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue).
#### Tests
- [ ] The code is covered in an existing or new test case using Google Test.
- [ ] The code coverage of the project (reported in the CI) should not decrease. If coverage is decreased, make sure that this is reasonable and acceptable.
- [ ] Valgrind doesn't find any bugs in the new code. [This script](https://github.com/DLR-AMR/t8code/blob/main/scripts/check_valgrind.sh) can be used to check for errors; see also this [wiki article](https://github.com/DLR-AMR/t8code/wiki/Debugging-with-valgrind).

If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually).
#### Scripts and Wiki
- [ ] If a new directory with source files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example or tutorial and a Wiki article.
#### License
- [x] The author added a BSD statement to `doc/` (or already has one).

spenke91 commented on src/t8_forest/t8_forest_partition.h:
```diff
@@ -36,11 +36,14 @@ T8_EXTERN_C_BEGIN ();
 */
 void
-t8_forest_partition (t8_forest_t forest);
+t8_forest_partition (t8_forest_t forest, t8_weight_fcn_t *weight_callback);
```
Does it make sense to add the `weight_callback` as an extra argument here? To me, it seems like an unnecessary duplication, given that the forest already has the new member `weight_function`. 
My suggestion would be to omit the extra argument here and then use `forest->weight_function` within `t8_forest_partition` rather than the function argument? Otherwise my impression is that we are bound to sooner or later run into hard-to-debug problems of the argument and the member being out of sync.

What do you think?

spenke91 replied:
Does it make sense to add the `weight_callback` as an extra argument here? To me, it seems like an unnecessary duplication, given that the forest already has the new member `weight_function`. 
My suggestion would be to omit the extra argument here and then use `forest->weight_function` within `t8_forest_partition` rather than the function argument? Otherwise my impression is that we are bound to sooner or later run into hard-to-debug problems of the argument and the member being out of sync.

What do you think?

```suggestion
t8_forest_partition (t8_forest_t forest);
```

spenke91 commented on src/t8_forest/t8_forest_partition.cxx:
```diff
@@ -430,49 +430,101 @@ t8_forest_partition_create_tree_offsets (t8_forest_t forest)
+    return;
+  }
+
+  if (not weight_fcn) {
```
```suggestion
  // If no weight function is provided, compute the offsets solely based on the number of elements.
  if (not weight_fcn) {
```

spenke91 commented on src/t8_forest/t8_forest_partition.cxx:
```diff
@@ -430,49 +430,101 @@ t8_forest_partition_create_tree_offsets (t8_forest_t forest)
+    return local_sum;
+  }();
+
+  double const partition_weight_offset = [&] () {  // partial sum of the partition weights, excluding the local rank
```
```suggestion
  double const partition_weight_offset = [&] () {  // partial sum of the partition weights of all lower-rank processes (excluding the local rank)
```

dutkalex commented on test/t8_forest/t8_gtest_forest_commit.cxx:
```diff
@@ -114,32 +114,29 @@ t8_test_forest_commit_abp (t8_forest_t forest, int maxlevel)
 
-  /* partition the forest */
+  /* partition the forest with a unit weight function */
+  t8_forest_t forest_partition;
```
I agree that this is not the best approach to test this feature, and adding a dedicated test suite for the partitioning algorithm is the right way to do things. However, beyond comparing the result of the unweighted algorithm with the weighted one with a constant weight callback, I don't know how to test this in a robust way. For development purposes, I have used a lot of intrusive whitebox testing manually, but this would not be very helpful in a proper automated test suite...

spenke91 replied:
Ok, then I suggest you leave the test case as it currently is in this PR and I will move the weighted partitioning to another test case in a later PR.

spenke91 commented on src/t8_forest/t8_forest_general.h:
```diff
@@ -322,7 +330,8 @@ t8_forest_get_user_function (const t8_forest_t forest);
 void
-t8_forest_set_partition (t8_forest_t forest, const t8_forest_t set_from, int set_for_coarsening);
+t8_forest_set_partition (t8_forest_t forest, const t8_forest_t set_from, int set_for_coarsening,
+                         t8_weight_fcn_t *weight_callback);
```
As touched upon above, you are right that this strategy does not scale well with an increasing number of options. For this specific feature, however, we still think it would be preferable to go along with it, because it is so simple.

dutkalex commented:
Hi @spenke91!
I found some time to come back to this PR and make the changes you requested (i.e. rollback the breaking API change in favor of an additional function to set the weights callback, and rollback the test hack). I also rebased the PR on the head of main, and fixed the conflicts with the custom offsets feature that recently got merged. IIRC, you implemented this feature, so please let me know if you have questions/comments/complaints on the refactoring.
Btw, feel free to ignore this message. I hope you're having a good time during these winter holidays.
I wish you - and the t8code team - all the best for the new year!

dutkalex commented:
> @dutkalex I gathered some opinions from the other developers and got rid of the extra flag as you suggested 👍

Ok, perfect then 😊

Davknapp commented on test/t8_forest/t8_gtest_weighted_partitioning.cxx:
```diff
@@ -0,0 +1,232 @@
+  t8_debugf ("Create uniform base forest.\n");
+
+  // Initial uniform refinement level
+  const int level = 2;
```
What is the current run-time of the test? Maybe we can allow a higher-level for the production-mode.
